src/webgpu/shader/execution/robust_access.spec.ts

```diff
@@ -0,0 +1,426 @@
+    /** Undefined if no buffer binding is needed */
+    let bufferBindingSize: number | undefined = undefined;
+
+    // Declare the data that will be accessed to check robust access, as a buffer or a struct
+    // in the global scope or inside the test function itself.
+    const structDecl = `
+      struct S {
+        startCanary: array<u32, 10>;
+        data: ${type};
+        endCanary: array<u32, 10>;
+      };`;
+
+    switch (storageClass) {
+      case 'uniform':
+      case 'storage':
+        {
+          assert(_typeInfo.layout !== undefined);
+          const layout = _typeInfo.layout;
+          bufferBindingSize = layout.size;
+          const qualifiers = storageClass === 'storage' ? `storage, ${storageMode}` : storageClass;
+          globalSource += `
+          [[block]] struct TestData {
+            data: ${type};
+          };
+          [[group(0), binding(0)]] var<${qualifiers}> s: TestData;`;
+        }
+        break;
+
+      case 'private':
+      case 'workgroup':
+        usesCanary = true;
+        globalSource += structDecl;
+        globalSource += `var<${storageClass}> s: S;`;
+        break;
+
+      case 'function':
+        usesCanary = true;
+        globalSource += structDecl;
+        testFunctionSource += 'var s: S;';
+        break;
+    }
+
+    // Build the test function that will do the tests.
+
+    // If we use a local canary declared in the shader, initialize it.
+    if (usesCanary) {
+      testFunctionSource += `
+        for (var i = 0u; i < 10u; i = i + 1u) {
+          s.startCanary[i] = 0xFFFFFFFFu;
+          s.endCanary[i] = 0xFFFFFFFFu;
+        }`;
+    }
+
+    const indicesToTest = t.params.indexSigned
+      ? [
+          // Exactly in bounds (should be OK)
+          '0',
+          `${_typeInfo.arrayLength} - 1`,
+          // Exactly out of bounds
+          '-1',
+          `${_typeInfo.arrayLength}`,
+          // Far out of bounds
+          '-1000000',
+          '1000000',
+          `${kMinI32}`,
+          `${kMaxI32}`,
+        ]
+      : [
+          // Exactly in bounds (should be OK)
+          '0u',
+          `${_typeInfo.arrayLength}u - 1u`,
+          // Exactly out of bounds
+          `${_typeInfo.arrayLength}u`,
+          // Far out of bounds
+          '1000000u',
+          `${kMaxU32}u`,
+          `${kMaxI32}u`,
+        ];
+    const constZero = t.params.indexUniform ? 'constants.zero' : '0u';
+    const constZeroIndex = t.params.indexSigned ? `i32(${constZero})` : constZero;
+
+    /** Returns a different number each time, kind of like a `__LINE__` to ID the failing check. */
+    const nextErrorReturnValue = (() => {
+      let errorReturnValue = 0x1000;
+      return () => {
+        ++errorReturnValue;
+        return `0x${errorReturnValue.toString(16)}u`;
+      };
+    })();
+
+    // Produce the accesses to the variable.
+    for (const indexToTest of indicesToTest) {
+      // TODO check with constants too if WGSL allows it.
```

Comment:
I think it meant constexpr but I'm not sure. Could be overridable constants too, that would make sense as well.